src/room_member.cpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 Tunnler Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <thread>
+#include <mutex>
+
+#include "tunnler/tunnler.h"
+#include "tunnler/room.h"
+#include "tunnler/room_member.h"
+#include "tunnler/room_message_types.h"
+
+#include "tunnler/assert.h"
+
+#include "BitStream.h"
+#include "RakNetTypes.h"
+
+/// A special MAC address that tells the room we're joining to assign us a MAC address automatically.
+static const MacAddress NoPreferredMac = { 0, 0, 0, 0, 0, 0 };
+
+RoomMember::RoomMember() {
+    peer = RakNet::RakPeerInterface::GetInstance();
+}
+
+RoomMember::~RoomMember() {
+    if (receive_thread)
+        receive_thread->join();
+    RakNet::RakPeerInterface::DestroyInstance(peer);
+}
+
+void RoomMember::HandleWifiPackets(const RakNet::Packet* packet) {
+    RakNet::BitStream stream(packet->data, packet->length, false);
+
+    // Ignore the first byte, which is the message id.
+    stream.IgnoreBytes(sizeof(RakNet::MessageID));
+
+    // Parse the WifiPacket from the BitStream
+    uint8_t frame_type;
+    stream.Read(frame_type);
+    WifiPacket::PacketType type = static_cast<WifiPacket::PacketType>(frame_type);
+
+    WifiPacket wifi_packet{};
+    wifi_packet.type = type;
+    stream.Read(wifi_packet.channel);
+    stream.Read(wifi_packet.transmitter_address);
+    stream.Read(wifi_packet.destination_address);
+
+    uint32_t data_length;
+    stream.Read(data_length);
+
+    std::vector<uint8_t> data(data_length);
+    stream.Read(reinterpret_cast<char*>(data.data()), data_length);
+
+    wifi_packet.data = std::move(data);
+
+    if (type == WifiPacket::PacketType::Beacon) {
+        std::lock_guard<std::mutex> lock(beacon_mutex);
+        beacon_queue.emplace_back(std::move(wifi_packet));
+
+        // If we have more than the max number of buffered beacons, discard the oldest one
+        if (beacon_queue.size() > MaxBeaconQueueSize)
+            beacon_queue.pop_front();
+    } else {
+        // TODO(Subv): Add to ringbuffer for data / non-beacons
```

Comment:
that's imo too much work for little payoff, we can just pop_front() the data queue like we do for the beacon queue